### PR TITLE
IGNITE-18738 Cleanup indexes on transaction abort

### PR DESCRIPTION
**Pull request by SammyVimes**
https://issues.apache.org/jira/browse/IGNITE-18738

### CODE REVIEW

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/IndexCleanupTest.java [hunk 1]
```diff
@@ -0,0 +1,425 @@
...
+
+    private static final UUID TX_ID = UUID.randomUUID();
+
+    private static final HybridClock CLOCK = new HybridClockImpl();
+
+    private TestHashIndexStorage pkInnerStorage;
+    private TestSortedIndexStorage sortedInnerStorage;
+    private TestHashIndexStorage hashInnerStorage;
+    private TestMvPartitionStorage storage;
+    private TestPartitionDataStorage dataStorage;
```
Comment (sashapolo) [inline modules/table/src/test/java/org/apache/ignite/internal/table/distributed/IndexCleanupTest.java:99]: look like this field is not used

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/IndexCleanupTest.java [hunk 2]
```diff
@@ -0,0 +1,425 @@
...
+        var value = new TestValue(2, "bar");
+        BinaryRow tableRow = binaryRow(key, value);
+
+        writer.addWrite(storageUpdateHandler, rowUuid, tableRow);
+        writer.addWrite(storageUpdateHandler, rowUuid, null);
+
+        // Write intent is in the storage.
+        assertEquals(1, storage.rowsCount());
+
+        // But indexes are removed.
```
Comment (sashapolo) [inline modules/table/src/test/java/org/apache/ignite/internal/table/distributed/IndexCleanupTest.java:191]: Shall we check that they were not empty after the first write?

File: modules/table/src/main/java/org/apache/ignite/internal/table/distributed/StorageUpdateHandler.java [hunk 3]
```diff
@@ -133,6 +147,117 @@ public void handleUpdateAll(
...
+     * @param rowId Row id.
+     * @param previousValues Cursor with previous version of the row.
+     */
+    private void tryRemoveFromIndexes(BinaryRow rowToRemove, RowId rowId, Cursor<ReadResult> previousValues) {
+        TableSchemaAwareIndexStorage[] indexes = this.indexes.get().values().toArray(new TableSchemaAwareIndexStorage[0]);
+
+        ByteBuffer[] indexValues = new ByteBuffer[indexes.length];
+
+        // Precalculate value for every index.
+        for (int i = 0, indexesLength = indexes.length; i < indexesLength; i++) {
```
[the remark this answers is not in the document]
Comment (SammyVimes) [reply modules/table/src/main/java/org/apache/ignite/internal/table/distributed/StorageUpdateHandler.java:217]: Yes.

File: modules/table/src/test/java/org/apache/ignite/internal/table/distributed/IndexCleanupTest.java (hunk 2 above)
Comment (SammyVimes) [reply modules/table/src/test/java/org/apache/ignite/internal/table/distributed/IndexCleanupTest.java:191]: Right

**Review by sashapolo — APPROVED** [review]

**Pull request merged**